src/LoggedInRouter.js

```diff
@@ -7,11 +7,69 @@ import TeacherRouter from "./teacher/_routing/_TeacherRouter";
 import Settings from "./pages/Settings";
 import ArticleReader from "./reader/ArticleReader";
 import UserDashboard from "./userDashboard/UserDashboard";
+import React, { useState, useEffect } from "react";
+import ExtensionMessage from "./components/ExtensionMessage";
+import Feature from "../src/features/Feature";
+import LocalStorage from "./assorted/LocalStorage";
+import { isMobile } from "./utils/misc/mobileDetection";
+
+/*global chrome*/
 
 export default function LoggedInRouter({ api, setUser }) {
+  const [hasExtension, setHasExtension] = useState(false);
+  const [isChrome, setIsChrome] = useState(false);
+  const [open, setOpen] = useState(false);
+  const [displayedPopup, setDisplayedPopup] = useState(false);
+  useEffect(() => {
+    setDisplayedPopup(LocalStorage.displayedPopup());
+    let userAgent = navigator.userAgent;
+    if (userAgent.match(/chrome|chromium|crios/i) && isMobile() === false) {
```

Comment:
this is very convoluted code here.
could you try to disentangle it a bit? 
something like, extract the following two 
independent and cohesive functions, and use them 
inside the useEffect? 

```js

function runningInChrome() {
  let userAgent = navigator.userAgent;
  return userAgent.match(/chrome|chromium|crios/i) && isMobile() === false;
}

function extensionPresent(trueCallback, falseCallback) {
  if (chrome.runtime) {
    chrome.runtime.sendMessage(
      process.env.EXTENSION_ID,
      "You are on Zeeguu.org!",
      function (response) {
        if (chrome.runtime.lastError) {
          console.log(chrome.runtime.lastError);
        }
        if (response.message === true) {
          trueCallback();
        } else {
          falseCallback();
        }
      }
    );
  }
}
```

Reply:
I almost did what you suggested here. But the else statement is not placed correctly. The else belongs to "if(chrome.runtime)". I was not sure how to implement trueCallback() and falseCallback() in that case. And it does not seem that much shorter to do so. So, inside of the if else, I just set the states directly. Feel free to make changes to this though.